libs/wire-api/src/Wire/API/MLS/Keys.hs

```
@@ -76,8 +80,61 @@ instance ToSchema MLSPublicKey where
     }
+
+data JWK = JWK
+  { keyType :: String,
+    curve :: String,
+    pubX :: ByteString,
+    pubY :: Maybe ByteString
+  }
+  deriving (Show, Ord, Eq)
+
+instance ToSchema JWK where
+  schema =
+    object "JWK" $
+      JWK
+        <$> (.keyType) .= field "kty" schema
+        <*> (.curve) .= field "crv" schema
+        <*> (.pubX) .= field "x" base64URLSchema
+        <*> (.pubY) .= maybe_ (optField "y" base64URLSchema)
+
+type MLSPublicKeysJWK = MLSKeys JWK
+
+mlsKeysToPublicJWK ::
+  ( Member (ErrorS 'MLSNotEnabled) r
+  ) =>
+  MLSPrivateKeys ->
+  Sem r MLSPublicKeysJWK
+mlsKeysToPublicJWK (MLSPrivateKeys (_, ed) (_, ec256) (_, ec384) (_, ec521)) =
+  -- The kty parameter for ECDSA is "EC", for Ed25519 it's "OKP" (octet key
+  -- pair).
+  -- https://www.rfc-editor.org/rfc/rfc7518.html#section-6.1
+
+  -- The crv for P-256, P-384, and P-521 are their names.
+  -- https://www.rfc-editor.org/rfc/rfc7518.html#section-6.2.1.1
+
+  -- The x parameter is mandatory for all keys, the y parameter is mandatory for
+  -- all ECDSA keys.
+  -- https://www.rfc-editor.org/rfc/rfc7518.html#section-6.2.1
+  MLSKeys (JWK "OKP" "Ed25519" (BA.convert ed) Nothing)
+    <$> (uncurry (JWK "EC" "P-256") . second Just <$> splitXY (ECDSA.encodePublic (Proxy @Curve_P256R1) ec256))
+    <*> (uncurry (JWK "EC" "P-384") . second Just <$> splitXY (ECDSA.encodePublic (Proxy @Curve_P384R1) ec384))
+    <*> (uncurry (JWK "EC" "P-521") . second Just <$> splitXY (ECDSA.encodePublic (Proxy @Curve_P521R1) ec521))
+  where
+    -- Obtaining X and Y from an encoded curve point follows the logic of
+    -- Crypto.ECC's encodeECPoint and decodeECPoint (the module is not
+    -- exported). Points need to be encoded in uncompressed representation. This
+    -- is true for ECDSA.encodePublic.
+    -- https://www.rfc-editor.org/rfc/rfc8422#section-5.4.1
+    splitXY mxy = do
+      (m, xy) <- noteS @'MLSNotEnabled $ BA.uncons mxy
```

I would maybe throw an internal error here and in similar places.